crypto/engine/eng_int.h

```diff
@@ -45,6 +50,35 @@ extern CRYPTO_RWLOCK *global_engine_lock;
 
 # endif
 
+# define ENGINE_UP_STRUCT_REF(e, i)                             \
+    do {                                                        \
+        CRYPTO_UP_REF(&(e)->struct_ref, (i), (e)->lock);        \
+        engine_ref_debug((e), 0, 1);                            \
+    } while (0)
+
+# define ENGINE_DOWN_STRUCT_REF(e, i)                           \
+    do {                                                        \
+        CRYPTO_DOWN_REF(&(e)->struct_ref, (i), (e)->lock);      \
+        engine_ref_debug((e), 0, -1);                           \
+    } while (0)
+
+# define ENGINE_STRUCT_REF(e, i)                                \
+    CRYPTO_atomic_read(&(e)->struct_ref, (i), (e)->lock)
```

Comment:
Reference counters and reading the value do not work together, there is a reason the reference counters don't have a method to read the current counter. At some point it was added and reverted. Please rewrite things in such a way that you don't need this function (and NGINE_STRUCT_REF).

Reply:
How is the code going to check for a zero function reference count without reading the value?
If the value needs to be read, it should be read atomically.  I guess I could increment and then decrement it.

This macro is only used in one place (`ENGINE_ctrl` in _eng_ctrl.c_).  Is it possible for an engine to exist with a zero structural reference count?  The count starts at 1 and is only decremented by _engine_free_util_ which frees the structure.  Perhaps, removing this code from `ENGINE_ctrl` is better?

Reply:
You check the value down_ref returns, and if that returns 0, you know there are no references anymore.

Reply:
The code in ENGINE_ctrl isn't decrementing the reference count, it is checking it isn't zero.
I've removed this code since it doesn't look like it is possible to do an ENGINE_ctrl on a non-free structure with a zero reference count.

I've removed both of the _REF macros as well by switching the functional reference count to use atomic add directly (this one needs to be read safe as well).

Reply:
ENGINE_FUNCT_REF is still there ...

Reply:
Sorry, yes it is there.  A disconnect between my fingers and the code.

It isn't accessing a `CRYPTO_REF_COUNT`, rather it is accessing an `int`.  The unsupported read operation looks to be required for functional references.  Nothing is deallocated when the functional reference count is reduced to zero.  I.e. they are not really a reference count in the libcrypto context.

I'll take another look next week and am open to suggestions as to how to change these without introducing excessive overhead.

Reply:
I can't see a nice way to handle the functional references using atomic operations without a read being available.  I can revise this PR to remove the atomic operations on the functional references -- they were all safe before being properly guarded by the global lock.

Reply:
If we're going for these atomic operations, one sane way would be to regard each use of an ENGINE pointer in a function as a reference, and that entering such a function should start with incrementing the ref counter, and decrementing it at the end.  Incidently, that would probably replace the global lock entirely, except when handling the global lists (there's no getting away from a global lock with those)

Reply:
Note: I'm not entirely sure about what I just wrote.  I brain keeps going back and forth on the subject, which is also the reason why I keep revisiting this...